src/Analyser/NodeScopeResolver.php

```diff
@@ -1833,6 +1840,14 @@ function (MutatingScope $scope) use ($expr, $nodeCallback, $context): Expression
 					return $type;
 				});
 
+				$validCount = IntegerRangeType::fromInterval(0, null);
+				if ($validCount->isSuperTypeOf($decrementedCountType)->yes()) {
+					$scope = $scope->assignExpression(
+						$countExpr,
+						$decrementedCountType
+					);
+				}
+
```

Comment:
just an initial idea on how this could be implemented. I am not sure its the right way to fix it, so figured I would submit it for review and further discussion.

I already realized that with this PR applied the following does not yet work:
```
/** @param array<int, string> $headers */
function pop1(array $headers): void
{
	if (count($headers) >= 4) {
		assertType('int<4, max>', count($headers));
		array_pop($headers);
		assertType('int<3, max>', sizeof($headers));
	}
}

/** @param array<int, string> $headers */
function foo(array $headers): void
{
	if (sizeof($headers) >= 4) {
		\PHPStan\dumpType(sizeof($headers));
		array_pop($headers);
		\PHPStan\dumpType(sizeof($headers));
	}
}
```

could be fixed - I guess - in a similar way as we already do for `count` here. but maybe there is a more elegant way?